File: plugins/ct-fuzz/harness/go/targets.go
// Registered Go targets.
//
// Each target is f(public, secret []byte) and uses the inputs in some way.
// The harness times one call. Public input is fresh-random each call;
// secret is fixed (class A) or random (class B) each call.

package main

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/ecdsa"
	"crypto/ed25519"
	"crypto/elliptic"
	"crypto/mlkem"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/subtle"
	"math/big"

	"filippo.io/edwards25519"
	"golang.org/x/crypto/curve25519"
)

// sink prevents the optimizer from eliding work whose result is unused.
var sink uint64

func registerAll() {
	// ---------- Known constant-time (precision targets) ----------

	// crypto/subtle.ConstantTimeCompare on 32-byte buffers.
	register("subtle_ConstantTimeCompare_32", 32, 32, 2000, func(pub, sec []byte) {
		sink += uint64(subtle.ConstantTimeCompare(pub, sec))
	})

	// crypto/subtle.ConstantTimeCopy: cmov-style. dst is closure-captured to
	// keep the allocator out of the timed region.
	{
		dst := make([]byte, 32)
		register("subtle_ConstantTimeCopy_32", 32, 32, 1000, func(pub, sec []byte) {
			subtle.ConstantTimeCopy(int(sec[0]&1), dst, pub)
			sink += uint64(dst[0])
		})
	}

	// curve25519 X25519 scalar mult: documented constant-time.
	register("curve25519_X25519", 32, 32, 1, func(pub, sec []byte) {
		out, err := curve25519.X25519(sec, pub[:32])
		if err == nil {
			sink += uint64(out[0])
		}
	})

	// edwards25519 scalar mult on basepoint: constant-time.
	// Use SetUniformBytes (64-byte input, mod-L reduction) so both classes
	// always reach the scalar-mult — no rejection-sampling asymmetry.
	register("edwards25519_ScalarBaseMult", 64, 32, 1, func(pub, sec []byte) {
		scalar, err := edwards25519.NewScalar().SetUniformBytes(sec)
		if err != nil {
			return
		}
		p := edwards25519.NewIdentityPoint().ScalarBaseMult(scalar)
		bs := p.Bytes()
		sink += uint64(bs[0])
	})

	// ---------- Known variable-time (recall targets) ----------

	// bytes.Equal: early-exit byte comparison.
	register("bytes_Equal_32", 32, 32, 2000, func(pub, sec []byte) {
		if bytes.Equal(pub, sec) {
			sink++
		}
	})

	// Naive byte loop with early exit on first mismatch.
	register("naive_eq_32", 32, 32, 2000, func(pub, sec []byte) {
		eq := true
		for i := 0; i < len(pub) && i < len(sec); i++ {
			if pub[i] != sec[i] {
				eq = false
				break
			}
		}
		if eq {
			sink++
		}
	})

	// Table lookup indexed by secret byte (cache timing risk).
	register("table_lookup_secret_index", 8, 32, 2000, func(pub, sec []byte) {
		var table [256]uint64
		for i := range table {
			table[i] = uint64(i) * 0x9e3779b97f4a7c15
		}
		// Index a 256-entry table by the first secret byte.
		sink += table[sec[0]]
		_ = pub
	})

	// math/big.Int.Mod: division is variable-time.
	register("bigint_mod_secret", 32, 32, 5, func(pub, sec []byte) {
		x := new(big.Int).SetBytes(sec)
		m := new(big.Int).SetBytes(pub)
		if m.Sign() == 0 {
			return
		}
		r := new(big.Int).Mod(x, m)
		if len(r.Bits()) > 0 {
			sink += uint64(r.Bits()[0])
		}
	})

	// ---------- Borderline / production targets ----------

	// crypto/elliptic P-256 ScalarBaseMult. Modern Go uses the constant-time
	// nistec implementation; older Go versions did not.
	register("elliptic_P256_ScalarBaseMult", 32, 0, 1, func(pub, sec []byte) {
		_ = pub
		x, y := elliptic.P256().ScalarBaseMult(sec)
		if x != nil {
			sink += uint64(x.Bits()[0])
			_ = y
		}
	})

	// RSA decryption — two variants. With rng=nil Go SKIPS blinding (variable-
	// time, known leaky). With rng=cryptoRandReader Go applies blinding (CT-ish).
	rsaKey := loadRSAKey()
	rsaRand := cryptoRandReader{}
	// Class A: ciphertext = 0xAA-fill (pub buf). Class B: random ciphertext.
	// The "secret" buffer is the ciphertext — sec varies per class, pub stays fixed.
	register("rsa_decrypt_unblinded", 256, 128, 1, func(pub, sec []byte) {
		_ = pub
		_, _ = rsa.DecryptPKCS1v15(nil, rsaKey, sec)
	})
	register("rsa_decrypt_blinded", 256, 128, 1, func(pub, sec []byte) {
		_ = pub
		_, _ = rsa.DecryptPKCS1v15(rsaRand, rsaKey, sec)
	})

	// SHA-256 of secret: HMAC-style; should be constant-time in input length.
	register("sha256_secret", 32, 32, 200, func(pub, sec []byte) {
		_ = pub
		h := sha256.Sum256(sec)
		sink += uint64(h[0])
	})

	// ---------- Production targets: vary the KEY, fixed plaintext/message ----------
	// These are the public APIs production Go code calls for AEAD and signatures.
	// Threat model: attacker submits chosen plaintexts/messages (held fixed
	// across our test) and observes wall-clock latency; we ask whether timing
	// depends on the secret key.

	// AES-128-GCM seal — varying 16-byte AES key.
	{
		nonce := make([]byte, 12) // fixed all-zeros
		pt := bytes.Repeat([]byte{0xAA}, 64)
		dst := make([]byte, 0, 80) // 64 bytes pt + 16 bytes tag
		register("aes128gcm_seal_vary_key", 16, 0, 5, func(pub, sec []byte) {
			_ = pub
			block, err := aes.NewCipher(sec)
			if err != nil {
				return
			}
			gcm, err := cipher.NewGCM(block)
			if err != nil {
				return
			}
			ct := gcm.Seal(dst[:0], nonce, pt, nil)
			sink += uint64(ct[0])
		})
	}

	// AES-128-GCM open with bogus tag — varying key. Always fails authentication.
	// Tests whether the failure-path timing leaks the key (Bleichenbacher-class).
	{
		nonce := make([]byte, 12)
		bogusCt := bytes.Repeat([]byte{0xAA}, 80) // 64-byte payload + 16-byte tag
		dst := make([]byte, 0, 64)
		register("aes128gcm_open_invalid_vary_key", 16, 0, 5, func(pub, sec []byte) {
			_ = pub
			block, err := aes.NewCipher(sec)
			if err != nil {
				return
			}
			gcm, err := cipher.NewGCM(block)
			if err != nil {
				return
			}
			out, err := gcm.Open(dst[:0], nonce, bogusCt, nil)
			if err == nil && len(out) > 0 {
				sink += uint64(out[0])
			}
		})
	}

	// Ed25519 sign — varying 32-byte private seed. Sign expands the seed via
	// SHA-512 internally and runs scalar mult on the basepoint. Both should be CT.
	{
		msg := bytes.Repeat([]byte{0xAA}, 64)
		register("ed25519_sign_vary_key", 32, 0, 1, func(pub, sec []byte) {
			_ = pub
			priv := ed25519.NewKeyFromSeed(sec)
			sig := ed25519.Sign(priv, msg)
			sink += uint64(sig[0])
		})
	}

	// ECDSA P-256 sign — varying 32-byte private scalar. Modern Go (1.18+) uses
	// crypto/internal/nistec for constant-time scalar mult. Pre-1.18 was variable-
	// time and exploitable. SignASN1 requires PublicKey.X/Y so we derive via
	// ScalarBaseMult — this means each call performs two scalar mults (key
	// derivation + sign), modeling the key-import-then-sign pipeline.
	{
		hash := bytes.Repeat([]byte{0xAA}, 32)
		register("ecdsa_p256_sign_vary_key", 32, 0, 1, func(pub, sec []byte) {
			_ = pub
			// Clamp to ensure D in [1, N-1].
			var d [32]byte
			copy(d[:], sec)
			d[0] &= 0x7F
			if d == ([32]byte{}) {
				d[31] = 1
			}
			x, y := elliptic.P256().ScalarBaseMult(d[:])
			priv := &ecdsa.PrivateKey{
				PublicKey: ecdsa.PublicKey{Curve: elliptic.P256(), X: x, Y: y},
				D:         new(big.Int).SetBytes(d[:]),
			}
			sig, err := ecdsa.SignASN1(rand.Reader, priv, hash)
			if err == nil && len(sig) > 0 {
				sink += uint64(sig[0])
			}
		})
	}

	// ---------- Production targets: SPLIT (prep untimed, measure timed) ----------
	// The same operations as above, but with the per-call cipher/key construction
	// moved OUT of the timed window. Compare to the whole-pipeline variants to
	// see whether the dudect signal lives in the cipher core or in the scaffolding.

	// AES-128-GCM seal, prep/measure split: only the gcm.Seal call is timed.
	{
		nonce := make([]byte, 12)
		pt := bytes.Repeat([]byte{0xAA}, 64)
		dst := make([]byte, 0, 80)
		var gcmState cipher.AEAD
		prep := func(sec []byte) {
			block, err := aes.NewCipher(sec)
			if err != nil {
				return
			}
			gcmState, _ = cipher.NewGCM(block)
		}
		measure := func(pub []byte) {
			_ = pub
			if gcmState == nil {
				return
			}
			ct := gcmState.Seal(dst[:0], nonce, pt, nil)
			sink += uint64(ct[0])
		}
		registerSplit("aes128gcm_seal_vary_key_split", 16, 0, 5, prep, measure)
	}

	// AES-128-GCM open with bogus tag, split. Only gcm.Open is timed.
	{
		nonce := make([]byte, 12)
		bogusCt := bytes.Repeat([]byte{0xAA}, 80)
		dst := make([]byte, 0, 64)
		var gcmState cipher.AEAD
		prep := func(sec []byte) {
			block, err := aes.NewCipher(sec)
			if err != nil {
				return
			}
			gcmState, _ = cipher.NewGCM(block)
		}
		measure := func(pub []byte) {
			_ = pub
			if gcmState == nil {
				return
			}
			out, err := gcmState.Open(dst[:0], nonce, bogusCt, nil)
			if err == nil && len(out) > 0 {
				sink += uint64(out[0])
			}
		}
		registerSplit("aes128gcm_open_invalid_vary_key_split", 16, 0, 5, prep, measure)
	}

	// Ed25519 sign, split. Key seed expansion (SHA-512) is in prep; only the
	// scalar-mult-and-output sign step is timed.
	{
		msg := bytes.Repeat([]byte{0xAA}, 64)
		var privKey ed25519.PrivateKey
		prep := func(sec []byte) {
			privKey = ed25519.NewKeyFromSeed(sec)
		}
		measure := func(pub []byte) {
			_ = pub
			sig := ed25519.Sign(privKey, msg)
			sink += uint64(sig[0])
		}
		registerSplit("ed25519_sign_vary_key_split", 32, 0, 1, prep, measure)
	}

	// ECDSA P-256 sign, split. Key derivation (ScalarBaseMult to get
	// PublicKey.X/Y, big.Int.SetBytes) is in prep; only ecdsa.SignASN1 is timed.
	{
		hash := bytes.Repeat([]byte{0xAA}, 32)
		var privKey *ecdsa.PrivateKey
		prep := func(sec []byte) {
			var d [32]byte
			copy(d[:], sec)
			d[0] &= 0x7F
			if d == ([32]byte{}) {
				d[31] = 1
			}
			x, y := elliptic.P256().ScalarBaseMult(d[:])
			privKey = &ecdsa.PrivateKey{
				PublicKey: ecdsa.PublicKey{Curve: elliptic.P256(), X: x, Y: y},
				D:         new(big.Int).SetBytes(d[:]),
			}
		}
		measure := func(pub []byte) {
			_ = pub
			sig, err := ecdsa.SignASN1(rand.Reader, privKey, hash)
			if err == nil && len(sig) > 0 {
				sink += uint64(sig[0])
			}
		}
		registerSplit("ecdsa_p256_sign_vary_key_split", 32, 0, 1, prep, measure)
	}

	// golang.org/x/crypto/curve25519.X25519 — split version. Vary the scalar.
	// X25519(scalar, point) — fixed peer point, vary the scalar. Production
	// X25519 used by TLS 1.3 ECDHE.
	{
		peer := bytes.Repeat([]byte{0x09}, 32) // basepoint-like fixed peer
		var scalar [32]byte
		prep := func(sec []byte) {
			copy(scalar[:], sec)
		}
		measure := func(pub []byte) {
			_ = pub
			out, err := curve25519.X25519(scalar[:], peer)
			if err == nil && len(out) > 0 {
				sink += uint64(out[0])
			}
		}
		registerSplit("curve25519_X25519_split", 32, 0, 1, prep, measure)
	}

	// Go stdlib ML-KEM-768 Decapsulate — varying the private key. Production
	// post-quantum KEM. Threat model: attacker submits a fixed ciphertext to
	// victim, victim Decapsulates with their private key; timing leak reveals
	// key bits.
	{
		// Pre-build a valid encapsulation+ciphertext from a reference key so
		// our ciphertext input is well-formed (otherwise decap exercises
		// the implicit-rejection path which is also CT but a different test).
		refKey, _ := mlkem.GenerateKey768()
		_, ciphertext := refKey.EncapsulationKey().Encapsulate()
		_ = ciphertext
		var dk *mlkem.DecapsulationKey768
		// 64-byte secret = ML-KEM seed (d || z, 32+32 bytes).
		prep := func(sec []byte) {
			var seed [64]byte
			copy(seed[:], sec)
			k, err := mlkem.NewDecapsulationKey768(seed[:])
			if err == nil {
				dk = k
			}
		}
		measure := func(pub []byte) {
			_ = pub
			if dk == nil {
				return
			}
			ss, err := dk.Decapsulate(ciphertext)
			if err == nil && len(ss) > 0 {
				sink += uint64(ss[0])
			}
		}
		registerSplit("mlkem768_decapsulate_vary_key_split", 64, 0, 1, prep, measure)
	}

	// ---------- v5: SUB-REGION annotation. Time exactly one phase. ----------

	// AES-128-GCM key schedule only: aes.NewCipher(sec).
	{
		var stashed [16]byte
		prep := func(sec []byte) { copy(stashed[:], sec) }
		measure := func(pub []byte) {
			_ = pub
			block, err := aes.NewCipher(stashed[:])
			if err == nil {
				_ = block
				sink++
			}
		}
		registerSplit("aes128gcm_keysched_only", 16, 0, 50, prep, measure)
	}

	// AES-128-GCM NewGCM only: cipher.NewGCM(block) given a pre-built cipher.Block.
	{
		var blockState cipher.Block
		prep := func(sec []byte) {
			b, err := aes.NewCipher(sec)
			if err == nil {
				blockState = b
			}
		}
		measure := func(pub []byte) {
			_ = pub
			gcm, err := cipher.NewGCM(blockState)
			if err == nil {
				_ = gcm
				sink++
			}
		}
		registerSplit("aes128gcm_newgcm_only", 16, 0, 50, prep, measure)
	}
}

// loadRSAKey returns a fresh 2048-bit RSA key generated at startup.
func loadRSAKey() *rsa.PrivateKey {
	k, _ := rsa.GenerateKey(cryptoRandReader{}, 2048)
	return k
}

type cryptoRandReader struct{}

func (cryptoRandReader) Read(p []byte) (int, error) { return realRandRead(p) }

File: plugins/ct-fuzz/harness/rust/src/targets.rs
use crate::{MeasureFn, PrepFn, TargetFn, TargetSpec};
use std::cell::RefCell;
use std::sync::atomic::{AtomicU64, Ordering};

use aes::cipher::{BlockEncrypt, KeyInit};
use aes::Aes128;
use curve25519_dalek::constants::ED25519_BASEPOINT_TABLE;
use curve25519_dalek::scalar::Scalar;
use num_bigint::BigUint;
use num_traits::Zero;
use ring::aead::{Aad, LessSafeKey, Nonce, UnboundKey, AES_128_GCM, CHACHA20_POLY1305};
use ring::signature::Ed25519KeyPair;
use aes_gcm::aead::{Aead, KeyInit as AesGcmKeyInit};
use aes_gcm::Aes128Gcm;
use chacha20poly1305::ChaCha20Poly1305;
use ed25519_dalek::{Signer, SigningKey};
use sha2::{Digest, Sha256};
use subtle::ConstantTimeEq;

static SINK: AtomicU64 = AtomicU64::new(0);

fn touch(v: u64) {
    SINK.fetch_add(v, Ordering::Relaxed);
}

fn whole(name: &'static str, secret_len: usize, public_len: usize, inner: usize, f: TargetFn) -> (&'static str, TargetSpec) {
    (name, TargetSpec::whole(secret_len, public_len, inner, f))
}

fn split(name: &'static str, secret_len: usize, public_len: usize, inner: usize, prep: PrepFn, measure: MeasureFn) -> (&'static str, TargetSpec) {
    (name, TargetSpec::split(secret_len, public_len, inner, prep, measure))
}

pub fn register_all() -> Vec<(&'static str, TargetSpec)> {
    let mut v: Vec<(&'static str, TargetSpec)> = Vec::new();

    // ---------- Known constant-time ----------

    v.push(whole("subtle_ConstantTimeEq_32", 32, 32, 2000, Box::new(|pub_, sec| {
        let r = pub_.ct_eq(sec);
        touch(r.unwrap_u8() as u64);
    })));

    v.push(whole("curve25519_dalek_scalar_mul", 32, 32, 1, Box::new(|_pub, sec| {
        let mut k = [0u8; 32];
        k.copy_from_slice(sec);
        let s = Scalar::from_bytes_mod_order(k);
        let p = &s * ED25519_BASEPOINT_TABLE;
        touch(p.compress().to_bytes()[0] as u64);
    })));

    v.push(whole("aes_block_encrypt", 16, 16, 200, Box::new(|pub_, sec| {
        let key = aes::cipher::generic_array::GenericArray::from_slice(sec);
        let cipher = Aes128::new(key);
        let mut block = aes::cipher::generic_array::GenericArray::clone_from_slice(pub_);
        cipher.encrypt_block(&mut block);
        touch(block[0] as u64);
    })));

    v.push(whole("sha2_256_secret", 32, 32, 200, Box::new(|_pub, sec| {
        let h = Sha256::digest(sec);
        touch(h[0] as u64);
    })));

    // ---------- Known variable-time ----------

    v.push(whole("naive_eq_32", 32, 32, 2000, Box::new(|pub_, sec| {
        let mut eq = true;
        for i in 0..pub_.len().min(sec.len()) {
            if pub_[i] != sec[i] {
                eq = false;
                break;
            }
        }
        if eq {
            touch(1);
        }
    })));

    v.push(whole("naive_pkeq_eq", 32, 32, 2000, Box::new(|pub_, sec| {
        if pub_ == sec {
            touch(1);
        }
    })));

    v.push(whole("table_lookup_secret_index", 8, 32, 2000, Box::new(|_pub, sec| {
        let mut table = [0u64; 256];
        for i in 0..256 {
            table[i] = (i as u64).wrapping_mul(0x9e3779b97f4a7c15);
        }
        touch(table[sec[0] as usize]);
    })));

    v.push(whole("num_bigint_mod_secret", 32, 32, 5, Box::new(|pub_, sec| {
        let x = BigUint::from_bytes_be(sec);
        let m = BigUint::from_bytes_be(pub_);
        if m.is_zero() {
            return;
        }
        let r = x % m;
        let bytes = r.to_bytes_be();
        if let Some(b) = bytes.first() {
            touch(*b as u64);
        }
    })));

    // ---------- Borderline / production ----------

    v.push(whole("ring_constant_time_verify", 32, 32, 2000, Box::new(|pub_, sec| {
        let _ = ring::constant_time::verify_slices_are_equal(pub_, sec);
        touch(0);
    })));

    // ---------- Production: ring AEAD/Ed25519, vary the key, WHOLE pipeline ----------

    v.push(whole("ring_aes128gcm_seal_vary_key", 16, 0, 5, Box::new(|_pub, sec| {
        let unbound = match UnboundKey::new(&AES_128_GCM, sec) {
            Ok(k) => k,
            Err(_) => return,
        };
        let key = LessSafeKey::new(unbound);
        let nonce = Nonce::assume_unique_for_key([0u8; 12]);
        let mut buf = [0xAAu8; 64];
        if key.seal_in_place_separate_tag(nonce, Aad::empty(), &mut buf).is_ok() {
            touch(buf[0] as u64);
        }
    })));

    v.push(whole("ring_aes128gcm_open_invalid_vary_key", 16, 0, 5, Box::new(|_pub, sec| {
        let unbound = match UnboundKey::new(&AES_128_GCM, sec) {
            Ok(k) => k,
            Err(_) => return,
        };
        let key = LessSafeKey::new(unbound);
        let nonce = Nonce::assume_unique_for_key([0u8; 12]);
        let mut buf = [0xAAu8; 80];
        let _ = key.open_in_place(nonce, Aad::empty(), &mut buf);
        touch(buf[0] as u64);
    })));

    v.push(whole("ring_ed25519_sign_vary_key", 32, 0, 1, Box::new(|_pub, sec| {
        let kp = match Ed25519KeyPair::from_seed_unchecked(sec) {
            Ok(k) => k,
            Err(_) => return,
        };
        let msg = [0xAAu8; 64];
        let sig = kp.sign(&msg);
        touch(sig.as_ref()[0] as u64);
    })));

    // ---------- Production: SPLIT mode (prep untimed, measure timed) ----------
    // Compare to whole-pipeline targets above to attribute the dudect signal.

    // ring AES-128-GCM seal — split. Key construction in prep, only seal timed.
    {
        let cipher: RefCell<Option<LessSafeKey>> = RefCell::new(None);
        let prep: PrepFn = Box::new(move |sec: &[u8]| {
            let unbound = match UnboundKey::new(&AES_128_GCM, sec) {
                Ok(k) => k,
                Err(_) => return,
            };
            *cipher.borrow_mut() = Some(LessSafeKey::new(unbound));
        });
        // We need cipher again for measure — but it was moved into prep.
        // Use a separate Rc<RefCell<>> to share between prep and measure.
        let _ = prep; // discard — use the shared variant below
    }

    {
        use std::rc::Rc;
        let cipher: Rc<RefCell<Option<LessSafeKey>>> = Rc::new(RefCell::new(None));
        let cipher_p = cipher.clone();
        let prep: PrepFn = Box::new(move |sec: &[u8]| {
            let unbound = match UnboundKey::new(&AES_128_GCM, sec) {
                Ok(k) => k,
                Err(_) => return,
            };
            *cipher_p.borrow_mut() = Some(LessSafeKey::new(unbound));
        });
        let cipher_m = cipher.clone();
        let measure: MeasureFn = Box::new(move |_pub: &[u8]| {
            let c = cipher_m.borrow();
            let key = match c.as_ref() {
                Some(k) => k,
                None => return,
            };
            let nonce = Nonce::assume_unique_for_key([0u8; 12]);
            let mut buf = [0xAAu8; 64];
            if key.seal_in_place_separate_tag(nonce, Aad::empty(), &mut buf).is_ok() {
                touch(buf[0] as u64);
            }
        });
        v.push(split("ring_aes128gcm_seal_vary_key_split", 16, 0, 5, prep, measure));
    }

    // ring AES-128-GCM open invalid — split.
    {
        use std::rc::Rc;
        let cipher: Rc<RefCell<Option<LessSafeKey>>> = Rc::new(RefCell::new(None));
        let cipher_p = cipher.clone();
        let prep: PrepFn = Box::new(move |sec: &[u8]| {
            let unbound = match UnboundKey::new(&AES_128_GCM, sec) {
                Ok(k) => k,
                Err(_) => return,
            };
            *cipher_p.borrow_mut() = Some(LessSafeKey::new(unbound));
        });
        let cipher_m = cipher.clone();
        let measure: MeasureFn = Box::new(move |_pub: &[u8]| {
            let c = cipher_m.borrow();
            let key = match c.as_ref() {
                Some(k) => k,
                None => return,
            };
            let nonce = Nonce::assume_unique_for_key([0u8; 12]);
            let mut buf = [0xAAu8; 80];
            let _ = key.open_in_place(nonce, Aad::empty(), &mut buf);
            touch(buf[0] as u64);
        });
        v.push(split("ring_aes128gcm_open_invalid_vary_key_split", 16, 0, 5, prep, measure));
    }

    // ring Ed25519 sign — split. KeyPair construction (seed expansion via SHA-512)
    // happens in prep; only the actual sign call is timed.
    {
        use std::rc::Rc;
        let kp: Rc<RefCell<Option<Ed25519KeyPair>>> = Rc::new(RefCell::new(None));
        let kp_p = kp.clone();
        let prep: PrepFn = Box::new(move |sec: &[u8]| {
            *kp_p.borrow_mut() = Ed25519KeyPair::from_seed_unchecked(sec).ok();
        });
        let kp_m = kp.clone();
        let measure: MeasureFn = Box::new(move |_pub: &[u8]| {
            let k = kp_m.borrow();
            if let Some(kp) = k.as_ref() {
                let msg = [0xAAu8; 64];
                let sig = kp.sign(&msg);
                touch(sig.as_ref()[0] as u64);
            }
        });
        v.push(split("ring_ed25519_sign_vary_key_split", 32, 0, 1, prep, measure));
    }

    // ---------- v5: SUB-REGION annotation. Same target, different timed window. ----------
    // Each variant times exactly one phase of the AES-128-GCM seal pipeline so we
    // can attribute where the original whole-pipeline |t| signal came from.

    // KEYSCHED-ONLY: time only UnboundKey::new (the AES key schedule).
    {
        use std::rc::Rc;
        let stashed: Rc<RefCell<[u8; 16]>> = Rc::new(RefCell::new([0u8; 16]));
        let stashed_p = stashed.clone();
        let prep: PrepFn = Box::new(move |sec: &[u8]| {
            stashed_p.borrow_mut().copy_from_slice(sec);
        });
        let stashed_m = stashed.clone();
        let measure: MeasureFn = Box::new(move |_pub: &[u8]| {
            let s = stashed_m.borrow();
            let r = UnboundKey::new(&AES_128_GCM, &s[..]);
            // Force the result to materialize via touch — and drop happens in scope.
            if let Ok(u) = r {
                let _ = u; // dropped here
                touch(1);
            }
        });
        v.push(split("ring_aes128gcm_keysched_only", 16, 0, 5, prep, measure));
    }

    // DROP-ONLY: time only the LessSafeKey drop (zeroize round keys, etc.).
    {
        use std::rc::Rc;
        let cipher: Rc<RefCell<Option<LessSafeKey>>> = Rc::new(RefCell::new(None));
        let cipher_p = cipher.clone();
        let prep: PrepFn = Box::new(move |sec: &[u8]| {
            if let Ok(unbound) = UnboundKey::new(&AES_128_GCM, sec) {
                *cipher_p.borrow_mut() = Some(LessSafeKey::new(unbound));
            }
        });
        let cipher_m = cipher.clone();
        let measure: MeasureFn = Box::new(move |_pub: &[u8]| {
            let _dropped = cipher_m.borrow_mut().take(); // drop happens at end of scope
            touch(1);
        });
        v.push(split("ring_aes128gcm_drop_only", 16, 0, 5, prep, measure));
    }

    // SEAL-ONLY is already registered as ring_aes128gcm_seal_vary_key_split.

    // ---------- v6 / top-10 additions: more production AEADs and signatures ----------

    // ring ChaCha20-Poly1305 seal — varying 32-byte key.
    v.push(whole("ring_chacha20poly1305_seal_vary_key", 32, 0, 5, Box::new(|_pub, sec| {
        let unbound = match UnboundKey::new(&CHACHA20_POLY1305, sec) {
            Ok(k) => k,
            Err(_) => return,
        };
        let key = LessSafeKey::new(unbound);
        let nonce = Nonce::assume_unique_for_key([0u8; 12]);
        let mut buf = [0xAAu8; 64];
        if key.seal_in_place_separate_tag(nonce, Aad::empty(), &mut buf).is_ok() {
            touch(buf[0] as u64);
        }
    })));

    // ring ChaCha20-Poly1305 seal split (only seal timed).
    {
        use std::rc::Rc;
        let key: Rc<RefCell<Option<LessSafeKey>>> = Rc::new(RefCell::new(None));
        let key_p = key.clone();
        let prep: PrepFn = Box::new(move |sec: &[u8]| {
            if let Ok(u) = UnboundKey::new(&CHACHA20_POLY1305, sec) {
                *key_p.borrow_mut() = Some(LessSafeKey::new(u));
            }
        });
        let key_m = key.clone();
        let measure: MeasureFn = Box::new(move |_pub: &[u8]| {
            let c = key_m.borrow();
            if let Some(k) = c.as_ref() {
                let nonce = Nonce::assume_unique_for_key([0u8; 12]);
                let mut buf = [0xAAu8; 64];
                if k.seal_in_place_separate_tag(nonce, Aad::empty(), &mut buf).is_ok() {
                    touch(buf[0] as u64);
                }
            }
        });
        v.push(split("ring_chacha20poly1305_seal_vary_key_split", 32, 0, 5, prep, measure));
    }

    // RustCrypto aes-gcm AES-128-GCM seal — varying 16-byte key.
    v.push(whole("rustcrypto_aes128gcm_seal_vary_key", 16, 0, 5, Box::new(|_pub, sec| {
        let key = aes_gcm::Key::<Aes128Gcm>::from_slice(sec);
        let cipher = Aes128Gcm::new(key);
        let nonce = aes_gcm::Nonce::from_slice(&[0u8; 12]);
        let pt = [0xAAu8; 64];
        if let Ok(ct) = cipher.encrypt(nonce, pt.as_ref()) {
            touch(ct[0] as u64);
        }
    })));

    // RustCrypto chacha20poly1305 seal — varying 32-byte key.
    v.push(whole("rustcrypto_chacha20poly1305_seal_vary_key", 32, 0, 5, Box::new(|_pub, sec| {
        let key = chacha20poly1305::Key::from_slice(sec);
        let cipher = ChaCha20Poly1305::new(key);
        let nonce = chacha20poly1305::Nonce::from_slice(&[0u8; 12]);
        let pt = [0xAAu8; 64];
        if let Ok(ct) = cipher.encrypt(nonce, pt.as_ref()) {
            touch(ct[0] as u64);
        }
    })));

    // ed25519-dalek sign — varying 32-byte private seed.
    v.push(whole("ed25519_dalek_sign_vary_key", 32, 0, 1, Box::new(|_pub, sec| {
        let mut s = [0u8; 32];
        s.copy_from_slice(sec);
        let kp = SigningKey::from_bytes(&s);
        let msg = [0xAAu8; 64];
        let sig = kp.sign(&msg);
        touch(sig.to_bytes()[0] as u64);
    })));

    v
}
